### PR TITLE
Generalized SingleVariableReturnMappingSolution

### PR DESCRIPTION
**Pull request by tophmatthews**
Ref #11239

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C [hunk 1]
```diff
@@ -135,21 +142,15 @@ SingleVariableReturnMappingSolution::internalSolve(const Real effective_trial_st
...
   Real scalar_increment = 0.0;
-  const Real min_permissible_scalar = 0.0;
+  const Real min_permissible_scalar = minimumPermissibleValue(effective_trial_stress);
   const Real max_permissible_scalar = maximumPermissibleValue(effective_trial_stress);
   Real scalar_upper_bound = max_permissible_scalar;
-  Real scalar_lower_bound = 0.0;
+  Real scalar_lower_bound = min_permissible_scalar;
   unsigned int it = 0;
 
-  if (effective_trial_stress == 0.0)
```
Comment (tophmatthews) [inline modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: @bwspenc This is the only place where effective_trial_stress is used in this class, and preventing complete removal. I wonder if it is even necessary to skip the iteration if stress is zero? Certainly in my models, even if stress is zero, I still need to run a newton iteration.

**Comment by tophmatthews** [comment]
ping @bwspenc for review

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 1 above)
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: @tophmatthews I believe we were having trouble with some models that divided by stress in the first timestep in the earlier model development.  With @bwspenc's changes in the iteration, we might not need to guard against it anymore.  I think the creep models are the most likely to demonstrate the problem, if it still exists

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Ah, that would make sense. I am guessing (maybe hoping) the tests would have picked up on this?

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: I can't remember if the tests did catch the problem, but I'm fairly certain the bison example problems did; would you be willing to run those example problems with your branch to check please?

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: sure. Remember which ones?

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: the tensor mechanics 2D rodlet with finite strain. thank you!

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: I'm guessing stress will only be zero during the initial timestep?

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: yes - we'd see failure in the material solve immediately. If your change gets the example problem to _t_step = 2, we should be good to go. Thank you for checking this case!

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Good that will save time! All of the tensor mechanics 2DRZ problems run the first timestep fine, so it seems this may have been fixed.

**Comment by tophmatthews** [comment]
ping @bwspenc , or is someone else more appropriate?

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 1 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: I'm good with removing this, but I'd like to also slightly re-work the logic around the call to `returnMappingSolve()` in `RadialReturnStressUpdate` that prevents a divide by zero as well.  I'd say let's change:
```
  // Use Newton iteration to determine the scalar effective inelastic strain increment
  Real scalar_effective_inelastic_strain = 0;
  returnMappingSolve(effective_trial_stress, scalar_effective_inelastic_strain, _console);

  if (scalar_effective_inelastic_strain != 0.0)
    inelastic_strain_increment = deviatoric_trial_stress *
                                 (1.5 * scalar_effective_inelastic_strain / effective_trial_stress);
  else
    inelastic_strain_increment.zero();
```
to:
```
  // Use Newton iteration to determine the scalar effective inelastic strain increment
  Real scalar_effective_inelastic_strain = 0;
  if (!MooseUtils::absoluteFuzzyEqual(effective_trial_stress, 0.0))
  {
    returnMappingSolve(effective_trial_stress, scalar_effective_inelastic_strain, _console);
    if (scalar_effective_inelastic_strain != 0.0)
      inelastic_strain_increment = deviatoric_trial_stress *
                                   (1.5 * scalar_effective_inelastic_strain / effective_trial_stress);
    else
      inelastic_strain_increment.zero();
  }
  else
    inelastic_strain_increment.zero();
```
...unless you really need to call into the Newton iteration once in your model.

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C [hunk 2]
```diff
@@ -350,7 +351,7 @@ bool
 SingleVariableReturnMappingSolution::converged(const Real residual, const Real reference)
 {
   return (std::abs(residual) <= _absolute_tolerance ||
-          (std::abs(residual) / reference) <= _relative_tolerance);
+          std::abs(residual / reference) <= _relative_tolerance);
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Was this change needed?  Seems like it's about the same since reference is always positive.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 1 above)
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Yeah, I need to call Newton iteration at least once. There are contributions to the strain that are not related to the stress. We can add a flag to force at least one iteration, although it seems cleaner how it is.

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C [hunk 3]
```diff
@@ -83,6 +83,13 @@ SingleVariableReturnMappingSolution::SingleVariableReturnMappingSolution(
   }
 }
 
+Real
+SingleVariableReturnMappingSolution::minimumPermissibleValue(
+    const Real /*effective_trial_stress*/) const
+{
+  return std::numeric_limits<Real>::min();
```
[the remark this answers is not in the document]
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Shoot, really this should be `- std::numeric_limits<Real>::max()`, or perhaps `std::numeric_limits<Real>::lowest()` is the same thing

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C [hunk 4]
```diff
@@ -103,11 +110,11 @@ SingleVariableReturnMappingSolution::returnMappingSolve(const Real effective_tri
     if (!internalSolve(effective_trial_stress, scalar, iter_output_ptr))
     {
       if (iter_output_ptr)
-        mooseError(iter_output_ptr->str());
+        throw MooseException(iter_output_ptr->str());
```
[the remark this answers is not in the document]
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Not in the currently implemented models. I have seen it with my own. Sometimes it does through a Petsc error along the lines `Object isn't in correct state` which is related to #10662, but usually it works. I figure it's a lot better to have this than killing the solution.

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 2 above)
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: That is true in your current implementations, but there are cases where reference can be negative if the scalar increment is negative, or there is a different reference residual than currently implemented. I could protect this using `std::abs()` in `computeReferenceResidual`, but it seems easier just to move this parentheses and always ensure the reference residual is being calculated.

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 1 above)
Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Oh I see, it still needs to protect the divide by zero...That's identified in #11295 as well

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Yeah, actually I'm good with your fix in `RadialReturnStressUpdate`. It moves the decision to do the newton solve back one step into the individual model. I have a corresponding model like `RadialReturnStressUpdate` where I can force the call. It'll also fix the divide by zero in #11295.

File: modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C (hunk 4 above)
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C]: Yeah, I'm fine with making the change.  In my experience, throwing an exception doesn't always result in cutting the time step like I'd like it to.  That's really a separate issue from this. The worst it can do is die with an error like it does now.

**Review by bwspenc — APPROVED** [review]
